common-lib/src/main/scala/com/gu/mediaservice/lib/cleanup/SupplierProcessors.scala

```diff
@@ -74,33 +75,109 @@ object AlamyParser extends ImageProcessor {
   }
 }
 
-object AllStarParser extends ImageProcessor {
-  val SlashAllstar = """(.+)/Allstar""".r
-  val AllstarSlash = """Allstar/(.+)""".r
+object AllStarParser extends CanonicalisingImageProcessor {
+  override def getAgencyName = "Allstar Picture Library"
+  override def getCanonicalName: String = "Allstar"
+
+  case object AllstarRegexResultExtractor extends RegexResultExtractor {
+    private val AllstarInSlashDelimitedString = "((.*)/)?(Allstar( Picture Library)?)(/(.*))?".r
```

Comment:
This API is fairly expansive, given that AFAICS we only care about the prefix/suffix. (Adding a new RegexResultExtractor will mean doing the capture group dance.) Given the two examples here, feels possible to write something with an API that just consumes the regex to find the agency, and uses that to find prefixes/suffixes. IIUC this would mean a single definition of the search regex, rather than a custom unapply method,  for each extractor.

But this is well explained and tested – more asking for my own curiosity, as I can't think of a case w/ prefix/suffix that wouldn't work with this smaller API – perhaps there is one, though?

Reply:
In part, you're correct.  We didn't need the unapply, which was clever but not clear.  We've replaced with an abstract method.  This is needed because the regexes are not necessarily the same between the two sub-classes, and may have different numbers of capture groups.

Reply:
Gotcha. My suggestion above assume the form `prefix/agency/suffix`. Out-of-band discussions have suggested that this may not always be the case (I can't see examples in the tests, but a) I may have missed and b) there may be some in future), and we'd need a more expansive  API to deal with that, so happy to leave as is.